x/wasm/internal/keeper/ibc.go

```diff
@@ -77,28 +80,176 @@ func (k Keeper) ClaimCapability(ctx sdk.Context, cap *capabilitytypes.Capability
+	if err != nil {
+		return nil, err
+	}
+
+	var sender sdk.AccAddress // we don't know the sender
+	params := cosmwasm.NewEnv(ctx, sender, nil, contractAddr)
+	params.IBC = &ibcInfo
+
+	querier := QueryHandler{
+		Ctx:     ctx,
+		Plugins: k.queryPlugins,
+	}
+
+	gas := gasForContract(ctx)
+	mock, ok := MockContracts[contractAddr.String()]
+	if !ok { // hack for testing without wasmer
+		panic("not supported")
+	}
+	res, gasUsed, execErr := mock.AcceptChannel(codeInfo.CodeHash, params, order, version, connectionHops, prefixStore, cosmwasmAPI, querier, ctx.GasMeter(), gas)
+	consumeGas(ctx, gasUsed)
+	if execErr != nil {
+		return nil, sdkerrors.Wrap(types.ErrExecuteFailed, execErr.Error())
+	}
+	if !res.Result { // todo: would it make more sense to let the contract return an error instead?
+		return nil, sdkerrors.Wrap(types.ErrInvalid, res.Reason)
+	}
+	return res.RestrictCounterpartyVersions, nil
+}
+
+func (k Keeper) OnRecvPacket(ctx sdk.Context, contractAddr sdk.AccAddress, payloadData []byte, ibcInfo cosmwasm.IBCInfo) ([]byte, error) {
+	codeInfo, prefixStore, err := k.contractInstance(ctx, contractAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	var sender sdk.AccAddress // we don't know the sender
+	params := cosmwasm.NewEnv(ctx, sender, nil, contractAddr)
+	params.IBC = &ibcInfo
+
+	querier := QueryHandler{
+		Ctx:     ctx,
+		Plugins: k.queryPlugins,
+	}
+
+	gas := gasForContract(ctx)
+	mock, ok := MockContracts[contractAddr.String()]
+	if !ok { // hack for testing without wasmer
+		panic("not supported")
+	}
+	res, gasUsed, execErr := mock.OnReceive(codeInfo.CodeHash, params, payloadData, prefixStore, cosmwasmAPI, querier, ctx.GasMeter(), gas)
+	consumeGas(ctx, gasUsed)
+	if execErr != nil {
+		return nil, sdkerrors.Wrap(types.ErrExecuteFailed, execErr.Error())
+	}
+
+	// emit all events from this contract itself
+	events := types.ParseEvents(res.Log, contractAddr)
+	ctx.EventManager().EmitEvents(events)
+
+	// hack: use sdk messages here for simplicity
+	for _, m := range res.Messages {
+		if err := k.messenger.handleSdkMessage(ctx, contractAddr, m); err != nil {
+			return nil, err
+		}
+	}
+	return res.Acknowledgement, nil
+}
+
+func (k Keeper) OnAckPacket(ctx sdk.Context, contractAddr sdk.AccAddress, payloadData []byte, acknowledgement []byte, ibcInfo cosmwasm.IBCInfo) error {
+	codeInfo, prefixStore, err := k.contractInstance(ctx, contractAddr)
+	if err != nil {
+		return err
+	}
+
+	var sender sdk.AccAddress // we don't know the sender
+	params := cosmwasm.NewEnv(ctx, sender, nil, contractAddr)
+	params.IBC = &ibcInfo
+
+	querier := QueryHandler{
+		Ctx:     ctx,
+		Plugins: k.queryPlugins,
+	}
+
+	gas := gasForContract(ctx)
+	mock, ok := MockContracts[contractAddr.String()] // hack for testing without wasmer
+	if !ok {
+		panic("not supported")
+	}
+	res, gasUsed, execErr := mock.OnAcknowledgement(codeInfo.CodeHash, params, payloadData, acknowledgement, prefixStore, cosmwasmAPI, querier, ctx.GasMeter(), gas)
```

Comment:
As above, we need to pass minimally the channelID here.

Also, I would love to be able to detect ack as success/error in our go code (in a fixed manner) before sending to the contract

Reply:
We also need some way to map back to the original packet that we sent. This is not just an unsolicited acknowledgement (like receive), but an acknowledgement that ties to one specific packet.

In ibc transfer they take the [packet metadata (port and channel) and use it to determine the refund](https://github.com/cosmos/cosmos-sdk/blob/master/x/ibc-transfer/keeper/relay.go#L207-L219)

Reply:
The original payload from the contract is passed in the `payloadData`. The contract has it's source data and the ack so that it can match it proper. 
(Assuming the data contains an unique identifier. This is also an area where the rust cosmwasm framework may provide some tooling for non natural identifiers)

Reply:
The sender's channel/port are in `IBCPacketInfo` and passed via Environment

Reply:
Ah... so payload data is read from the local store from the original input, and isn't something sent from the other side (that is acknowledgement)

Reply:
I have not looked deep enough through the IBC impl to find where the origin message comes from in the ack process but it is available to us in the IBC callbacks.